File: Backtracking/leetcode_47.cpp

// Problem link ===>> https://leetcode.com/problems/permutations-ii/description/

#include <bits/stdc++.h>
#define ll long long
using namespace std;

class Solution
{
public:
    std::vector<std::vector<int>> permuteUnique(std::vector<int> &nums)
    {
        std::set<std::vector<int>> uniquePermutations;
        std::vector<int> currentPermutation;
        std::vector<bool> used(nums.size(), false); // Track used elements

        std::sort(nums.begin(), nums.end()); // Important for handling duplicates

        backtrack(nums, used, currentPermutation, uniquePermutations);

        std::vector<std::vector<int>> result(uniquePermutations.begin(), uniquePermutations.end());
        return result;
    }

private:
    void backtrack(std::vector<int> &nums, std::vector<bool> &used, std::vector<int> &currentPermutation, std::set<std::vector<int>> &uniquePermutations)
    {
        if (currentPermutation.size() == nums.size())
        {
            uniquePermutations.insert(currentPermutation);
            return;
        }

        for (int i = 0; i < nums.size(); ++i)
        {
            if (used[i])
                continue; // Skip if already used

            currentPermutation.push_back(nums[i]);
            used[i] = true;
            backtrack(nums, used, currentPermutation, uniquePermutations);
            used[i] = false;               // Backtrack
            currentPermutation.pop_back(); // Backtrack
        }
    }
};

int main()
{
    Solution solution;
    std::vector<int> nums = {1, 1, 2};
    std::vector<std::vector<int>> permutations = solution.permuteUnique(nums);

    for (const auto &perm : permutations)
    {
        for (int val : perm)
        {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}

File: Backtracking/leetcode_47.py

# Problem link ===>>  https://leetcode.com/problems/permutations-ii/description/
# submission link == =>> https://leetcode.com/problems/permutations-ii/submissions/1493413743/


from typing import List
class List:
    def __init__(self):
        self.data = []

    def append(self, item):
        self.data.append(item)


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        """
        Generates all unique permutations of a list containing duplicates.
        """
        result = []
        nums.sort()  # Crucial for handling duplicates efficiently

        def backtrack(current_permutation, remaining_nums):
            if not remaining_nums:
                result.append(current_permutation.copy())
                return

            for i in range(len(remaining_nums)):
                # Skip duplicates: If the current element is the same as the previous one
                # AND the previous one hasn't been used yet in the current permutation, skip it.
                if i > 0 and remaining_nums[i] == remaining_nums[i - 1]:
                    continue

                backtrack(
                    current_permutation + [remaining_nums[i]],
                    remaining_nums[:i] + remaining_nums[i + 1:],
                )

        backtrack([], nums)
        return result


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 1, 2]
    unique_perms = sol.permuteUnique(nums)
    print("Unique permutations:")
    for perm in unique_perms:
        print(perm)

    nums = [1, 2, 3]
    unique_perms = sol.permuteUnique(nums)
    print("Unique permutations:")
    for perm in unique_perms:
        print(perm)
